### PR TITLE
Sync: add a new method, do_only_first_initial_sync

### PR DESCRIPTION
**Pull request by kbrown9**
#### Changes proposed in this Pull Request:

* Add a new method, `do_only_first_initial_sync`, which is used to trigger an initial full sync only when one hasn't already been performed. 
* This method will be useful for plugins that want to run an initial full sync on sites that haven't already had an initial sync.

#### Jetpack product discussion
* n/a

#### Does this pull request change what data or activity we track or use?
* no

#### Testing instructions:
1. Install and activate this branch.
2. Connect to WPCOM.
3. Check that the full sync status option is populated with data. The `started` and `finished` values should be  timestamps. `wp option get jetpack_sync_full_status`
4. Using `wp shell`, try to start an initial sync as shown below. This should return false because an initial sync has already been run.
`return Automattic\Jetpack\Sync\Actions::do_only_first_initial_sync();`
5. Clear the `jetpack_sync_full_status` option: `wp option delete jetpack_sync_full_status`
6. Using `wp shell`, try to start an initial sync. This should return `null` because the full sync was started.
`return Automattic\Jetpack\Sync\Actions::do_only_first_initial_sync();`

### CODE REVIEW

File: projects/packages/sync/tests/php/test-actions.php [hunk 1]
```diff
@@ -0,0 +1,95 @@
+<?php // phpcs:ignore WordPress.Files.FileName.InvalidClassFileName
```
Comment (kbrown9) [inline projects/packages/sync/tests/php/test-actions.php:1]: I'm not sure if I should have just added these tests to the existing Sync unit test files in the `plugins/jetpack/tests/php/sync` directory.

Comment (mdbitz) [reply projects/packages/sync/tests/php/test-actions.php:1]: Here is more appropriate, the existing tests in the jetpack folder eventually need to be split having unit tests in the package then only the integration level tests in jetpack.

File: projects/packages/sync/src/class-actions.php [hunk 2]
```diff
@@ -507,6 +505,35 @@ public static function do_initial_sync() {
...
+			return false;
+		}
+
+		// Don't start a new sync if a full sync is in progress.
+		if ( $full_sync_module->is_started() && ! $full_sync_module->is_finished() ) {
+			return false;
+		}
+
+		// Don't start a new sync if the site isn't registering and an initial sync was already started.
+		if ( ! doing_action( 'jetpack_site_registered' ) && $full_sync_module->is_started() ) {
```
Comment (mdbitz) [inline projects/packages/sync/src/class-actions.php:530]: @kbrown9 can you outline why this check on 'jetpack_site_registered' is added? Logically it does make sense but it is a departure from existing functionality where the action 'jetpack_user_authorized' can also trigger do_initial_sync.

Also won't this check on is_started cause issues with sites that delete Jetpack than re-install?  

Would it make sense to add flag of $strict or similar that would preserve existing flow and then add this if statement only if strict is true? This way WC Pay can call `do_initial_sync( true )` and get the flow they want but not change existing flows?

**Review by mdbitz — COMMENTED** [review]
Question on the additional action/is_started check

File: projects/packages/sync/src/class-actions.php (hunk 2 above)
Comment (kbrown9) [reply projects/packages/sync/src/class-actions.php]: I like your proposal to add a flag. I think an even better approach would be to introduce a new method, `Actions::do_only_first_initial_sync`, which consumer plugins could call when they want an initial sync run only when one has never been run before.

The `Actions::do_intial_sync` method is hooked to a few actions, and the `jetpack_site_registered` action passes a [few arguments](https://github.com/Automattic/jetpack/blob/1d03317f5a562093fe0c4214fdf27092b3d092f9/projects/packages/connection/src/class-manager.php#L1180) to the callback functions. We specify that `Actions::do_initial_sync` accepts 0 arguments from the action hook [here](https://github.com/Automattic/jetpack/blob/532ea71e3da11560fe4c4874457ea9201bab30de/projects/packages/sync/src/class-main.php#L35), but if that ever needed to change, I think we would run into problems.

I think introducing the new `Actions::do_only_first_initial_sync` method will be both clearer and safer than adding a flag to the existing `Actions::do_intial_sync` method. What do you think of this approach?

Comment (mdbitz) [reply projects/packages/sync/src/class-actions.php]: I agree a new method is cleaner and safer for usage. 👍 from me on this approach

**Review by mdbitz — APPROVED** [review]
Thanks for the great test instructions. Verified and now Approving.

**Pull request merged**